### PR TITLE
statistics: add tests for `analyze` with dynamic partition prune mode

### PR DESCRIPTION
**Pull request by rebelice**
### What problem does this PR solve?

1. add tests for `analyze` with dynamic partition prune mode
2. fix a bug when merging global `topN`

Tests 

- Unit test

Side effects

### Release note 

- No release note

### CODE REVIEW

File: go.sum [hunk 1]
```diff
@@ -850,6 +846,8 @@ honnef.co/go/tools v0.0.0-20190523083050-ea95bdfd59fc/go.mod h1:rf3lG4BRIbNafJWh
 honnef.co/go/tools v0.0.1-2019.2.3/go.mod h1:a3bituU0lyd329TUQxRnasdCoJDkEUEAqEt0JzvZhAg=
 honnef.co/go/tools v0.1.1 h1:EVDuO03OCZwpV2t/tLLxPmPiomagMoBOgfPt0FM+4IY=
 honnef.co/go/tools v0.1.1/go.mod h1:NgwopIslSNH47DimFoV78dnkksY2EFtX0ajyb3K/las=
+honnef.co/go/tools v0.1.2 h1:SMdYLJl312RXuxXziCCHhRsp/tvct9cGKey0yv95tZM=
+honnef.co/go/tools v0.1.2/go.mod h1:NgwopIslSNH47DimFoV78dnkksY2EFtX0ajyb3K/las=
```
Comment (qw4990) [inline go.sum:850]: Why do we change it?

File: statistics/cmsketch.go [hunk 2]
```diff
@@ -670,12 +670,14 @@ func MergeTopN(topNs []*TopN, n uint32) (*TopN, []TopNMeta) {
 	}
 
 	var finalTopN TopN
+	remain := n
 	finalTopN.TopN = make([]TopNMeta, 0, n)
 	popedTopNPair := make([]TopNMeta, 0, uint32(numTop)-n)
 	for value, cnt := range counter {
 		data := hack.Slice(string(value))
-		if n > 0 && cnt >= lastTopCnt {
+		if n > 0 && remain > 0 && cnt >= lastTopCnt {
```
Comment (qw4990) [inline statistics/cmsketch.go:678]: It seems `n > 0` can be removed.

Comment (qw4990) [inline statistics/cmsketch.go:678]: The condition `remain > 0` may make the result unstable when multiple items have the same `lastTopCnt`.
Could we also update the sort algorithm above to make it stable?

File: go.sum (hunk 1 above)
Comment (rebelice) [reply go.sum:850]: `go mod tidy` changes it.

File: statistics/cmsketch.go (hunk 2 above)
Comment (rebelice) [reply statistics/cmsketch.go:678]: Oh, you are right.

File: statistics/handle/handle_test.go [hunk 3]
```diff
@@ -1069,6 +1069,32 @@ func (s *testStatsSuite) TestMergeIdxHist(c *C) {
 	c.Assert(len(rows.Rows()), Equals, 2)
 }
 
+func (s *testStatsSuite) TestAnalyzeWithDynamicPartitionPruneMode(c *C) {
```
Comment (Reminiscent) [inline statistics/handle/handle_test.go:1072]: Test LGTM

File: statistics/cmsketch.go [hunk 4]
```diff
@@ -662,23 +662,40 @@ func MergeTopN(topNs []*TopN, n uint32) (*TopN, []TopNMeta) {
...
 	}
 
 	var finalTopN TopN
-	finalTopN.TopN = make([]TopNMeta, 0, n)
-	popedTopNPair := make([]TopNMeta, 0, uint32(numTop)-n)
+	finalTopN.TopN = make([]TopNMeta, 0, remain)
+	popedTopNPair := make([]TopNMeta, 0, uint32(numTop)-remain)
+	boundaryTopN := make([]TopNMeta, 0, numTop)
 	for value, cnt := range counter {
 		data := hack.Slice(string(value))
```
Comment (Reminiscent) [inline statistics/cmsketch.go:677]: We can check `remain` first. For example, `if remain < 0 {break}`. And use `if / else if / else` to check the relation between `cnt` and `lastTopCnt`;

**Review by Reminiscent — APPROVED** [review]
/LGTM

**Pull request merged**